.github/workflows/verify-success.yml

```diff
@@ -0,0 +1,155 @@
+---
+name: Verify Success reusable workflow
+
+# Use this reusable workflow as a job of workflow to check that
+# all jobs, including ones ran through a matrix, were successful.
+# This job can then be used as a required status check in the
+# repo's rulesets, that allows to change the required jobs or
+# the matrix values of a required job without needing to change
+# the rulesets settings. In the future, GitHub might have a
+# solution to this natively.
+
+# This reusable workflow has inputs to change what is required
+# to have this workflow pass. It handles the cases were there were
+# skipped jobs, and no successful jobs.
+
+# The jobs to check must set as the `needs` for the job calling this
+# reusable workflow. This also means that the job ids should be in the
+# same workflow file. The calling job must be set to always run to be
+# triggered when jobs are skipped or cancelled.
+# Then, set the `needs_context` input like:
+# `needs_context: ${{ toJson(needs) }}`
+
+# Example usage, as a job inside a workflow:
+# ```
+# jobs:
+#   a-job-id:
+#     ...
+#   another-job-id:
+#     ...
+#   some-job-success:
+#     name: Some Job Result
+#     needs:
+#       - a-job-id
+#       - another-job-id
+#     if: ${{ always() }}
+#     uses: ./.github/workflows/verify-success.yml
+#     with:
+#       needs_context: ${{ toJson(needs) }}
+# ```
+
+on:
+  workflow_call:
+    inputs:
+      needs_context:
+        type: string
+        required: true
+        # Can't escape the handlebars in the description
+        description:
+          In the calling job that defines all the needed jobs,
+          send `toJson(needs)` inside `$` followed by `{{ }}`
+      fail_if_failure:
+        type: boolean
+        default: true
+        description:
+          If true, this workflow will fail if any job from 'needs_context was
+          failed
+      fail_if_cancelled:
+        type: boolean
+        default: true
+        description:
+          If true, this workflow will fail if any job from 'needs_context' was
+          cancelled
+      fail_if_skipped:
+        type: boolean
+        default: false
+        description:
+          If true, this workflow will fail if any job from 'needs_context' was
+          skipped
+      require_success:
+        type: boolean
+        default: true
+        description:
+          If true, this workflow will fail if no job from 'needs_context' was
+          successful
+
+jobs:
+  verify-success:
+    name: Success
```

Comment:
Any alternative name suggestions? This text goes to the job result:

> Python Code Quality / Python Code Quality Result / Success

I think it might be confusing if you see:

> ❌ Python Code Quality / Python Code Quality Result / Success

"It is red, but it says _success_."

Reply:
I agree that a change would be better. I don't have an idea yet.

While the line we see below in the checks run is `Python Code Quality / Python Code Quality Result / Success`,
what we specify to require passing is only this part : `Python Code Quality Result / Success`
This part needs to have a stable name in time. The last slash appears because it was a reusable workflow (there is a limit on the depth that a reusable workflow can call a reusable workflow of 4 levels). My previous inline solution was a bit smaller, only had its name, but would have needed to be maintained at all places we wanted, and if the extra handling of edge cases were removed, there could be surprises in the future that I already knew how to handle at this moment.
Something that makes us keep the link between what workflow it relates to and the associated "success"-checking job is desirable.

For choosing the name: I didn't want to use "Passed", as the distinction isn't clear with skipped. I wanted it to be the smallest possible. If we remove the `name:` line, it would default to the job id `verify-success` if I recall correctly. It's not awful, but prevents us of changing things around in the future as it is easy with this to have another solution that would keep the same name.

Reply:
Let me a couple more days to think about this specific issue a little more

Reply:
I agree with the intuition around Passed versus Success. Overall, I can't tell much difference in:

> ✅ Python Code Quality Result / Passed
> ❌ Python Code Quality Result / Passed

> ✅ Python Code Quality Result / Verify Success
> ❌ Python Code Quality Result / Verify Success

> ✅ Python Code Quality Result / Success
> ❌ Python Code Quality Result / Success

Going back to _Success_, ✅ Success and ❌ Success read sort of like _Success_ and _NOT Success_ which is exactly what we want.

Reply:
I kept thinking about this, and didn't find a better naming yet. I still consider it is the best, as that reusable workflow's purpose is to generate either a successful result, or a failed result, depending on its inputs. It can't be skipped or cancelled, it acts like a logic gate.

Reply:
✅ Success and ❌ Success it is.